README.md
# Memory Board User Guide

## Basic information

    Memory board version:           2020-09-18
    
    ADC model:                      AD7367BRUZ
    DAC model:                      AD5725BRSZ-500RL7
    Switch model:                   MT8816AF1
    Socket model:                   

    Opal Kelly device model:        XEM6010-LX150
    FrontPanel driver version:      5.1.3
    Xilinx FPGA model:              Spartan-6, XC6SLX150-2fgg484
    ISE Design Suite version:       14.7

    FPGA core language:             Verilog
    API language:                   Python 3.7

## Software Installation

### 1. Install Python3.7

Install **Miniconda3 Windows 64-bit** [Official webpage](https://docs.conda.io/en/latest/miniconda.html).
>Note: Allow Miniconda to modify PATH variables.

Open Anaconda Command Prompt.

```
conda create -n py3.7 python=3.7
conda activate py3.7
conda install numpy
```

### 2. Install **FrontPanel** SDK [Google Drive](https://drive.google.com/file/d/1HM5w99bJSepEbRAgtagARoK4IIzPZ-vO/view?usp=sharing)

After installing FrontPanel SDK, you should add FrontPanel Python API to system environment `"Path"`. The path should contains "ok.py" and "_ok.pyd" file and be similar to:

    your_path_to/Opal Kelly/FrontPanelUSB/API/Python/3.7/x64

>Note: Install driver-only from [Opal Kelly](https://pins.opalkelly.com/downloads) if the above link is unavailable.

### 3. Download **Memboard** Python library from [Github](https://github.com/frazier-zh/memboard-python-api/archive/refs/heads/master.zip)

## Powering The Board

## Example

In this example, 

```Python
import memboard as mb
from memboard import wait, time, apply, measure, reset
import memboard.unit as u

import numpy as np

def run():
    top_pin_list = [29, 30, 31, 32, 33]
    bottom_pin = 83
    results = np.zeros((2, 5))

    reset('all')
    apply(pin=1, v=0.5 *u.V)
    apply(pin=84, v=0 *u.V)

    for i in range(5):
        results[0, i] = time()
        results[1, i] = measure(pin=bottom_pin, drive_pin=top_pin_list[i], v=0.1 *u.V)
        wait(1 *u.us)

    return results


with mb.connect('../verilog/src/top.bit'):
    mb.register(run)
    mb.execute(every=20 *u.us, total=1 *u.s, out='scan')
```

## API Reference

### `connect(path, debug)`

|Parameters|Type|Details
|---|---|---|
|path|str|Indicates path of FPGA binary file "top.bit"|
|debug|bool|Indicates debug mode|

### `add(func)`

|Parameters|Type|Details
|---|---|---|

### `execute()`

### `reset()`

### `apply()`

### `ground()`

### `measure()`

### `wait()`

### `time()`

## Datasheet

### 1. Socket and Pin Mapping

### 2. Execution Time

|Parts|Operation (ns)|Reset (ns)|
|---|---:|---:|
|ADC|1390|20|
|DAC|60|50|
|Switch|190|190|

>Note: Every operation has an additional 30ns processing time.

### 3. Connections of Switches

|Switch Group|Pin|ADC|DAC|
|---|---|---|---|
|Source|1-28|Channel 0|GND*, Channel 1|
|Gate|29-54|-|Channel 2|
|Drain|55-84|Channel 1|Channel 3|

**Channel 0 should always be set to 0V (GND).*

### 4. Non-idealities

1. DAC maximum glitch amplitude is 1V.
2. DAC maximum glitch time is 1000ns.
3. Switch settling time is 100ns.

## FPGA Programming Guide

Refer to the [guide](verilog/README.md).


CHANGELOG.md
# Changelog

## [Unreleased]

### Feature

- Add global error handling module, unify the error code
- FPGA code stability fix
- FPGA interface upgrade
- DAC check

## [0.2] - 2022-02-26

### Added

- "CHANGELOG.md" for recording changes
- "board" class for board operation

### Changed

- Renamed "operation.py" to "board.py"
- Integrated all board operation to "board.py"
- IntEnum for storing constant
- Functions no longer raise errors, all error and warning are handled through module logger

### Removed

- allow_auto() is no longer used for executing single instruction

### Fixed

## [0.1] - 2022-02-23

### Added

- The first release of memboard python API
- The first release of corrsponding FPGA hardware code

### Known Issues

- Complication in staring an interactive session and an automated execution
- Complication in result conversion
- ADC BUSY signal is occasionally stuck in high after conversion
- Inaccurate operation runtime
- Device control are bundled, causing unexpected waiting
- FPGA has no error return or time out
- Complication in ADC control

  

memboard/const.py
""" Device List Class
    Enumerate every device
"""
from enum import IntEnum

DESYNC_TIME = 100

PACKET_SIZE = 12

class TerminalType(IntEnum):
    Sense = 0
    Force = 1

class State(IntEnum):
    idle = 0
    reset = 1
    start = 2
    wait = 3
    clear = 4
    busy = 5
    read = 8

class OP(IntEnum):
    logic = 0
    jmp = 1
    call = 2
    ret = 3
    loop = 5
    adc = 8
    dac = 9
    switch = 10
    clr = 11

socket_mapping = {
    'A': [32, 30, 29, 27, 24, 23, 20, 17, 15, 14, 11],
    'B': [35, 33, 31, 28, 25, 18, 19, 16, 13, 12,  9],
    'C': [36, 34, -1, -1, 26, 22, 21, -1, -1, 10,  8],
    'D': [38, 37, -1, -1, -1, -1, -1, -1, -1,  7,  6],
    'E': [41, 40, 42, -1, -1, -1, -1, -1,  2,  4,  3],
    'F': [44, 39, 43, -1, -1, -1, -1, -1,  1, 84,  5],
    'G': [45, 46, 47, -1, -1, -1, -1, -1, 81, 82, 83],
    'H': [48, 49, -1, -1, -1, -1, -1, -1, -1, 79, 80],
    'J': [50, 52, -1, -1, 60, 64, 65, -1, -1, 76, 78],
    'K': [51, 54, 55, 58, 61, 63, 67, 70, 73, 75, 77],
    'L': [53, 56, 57, 59, 62, 68, 66, 69, 71, 72, 74]
}


def to_pin(loc: str):
    col = loc.upper()[0]
    row = int(loc[1:])-1
    if (col in socket_mapping) and (row in range(1, 11+1)):
        pin = socket_mapping[col][row]
        if pin == -1:
            __logger.error(f"Socket-{loc} is unavailable.")
    else:
        __logger.error(f"Invalid socket location {loc}.")
    return pin

def to_socket(pin: int):
    for key, value in socket_mapping.items():
        if pin in value:
            return f'{key}{value}'

#   switch  GND     DAC     ADC
switch_connection = {
    0:     [0,      2,      0],
    1:     [0,      2,      0],
    2:     [-1,     1,     -1],
    3:     [-1,     1,     -1],
    4:     [-1,     3,      1],
    5:     [-1,     3,      1]
}

def get_channel(b: Base, sw_no):
    if b == Base.adc:
        return switch_connection[sw_no][2], 2
    elif b == Base.dac:
        return switch_connection[sw_no][1], 1

def to_switch(pin):
    if pin not in range(1, 84+1):
        __logger.error(f'Invalid pin number {pin}.')

    group = int((pin-1)/28)*2 + int((pin-1)%28/16)
    pin_no = (pin-1)%28%16
    return group, pin_no
    
class Port:
    def __init__(self, pin=None, socket=None, sw_no=None, sw_x=None):
        if pin is not None:
            self.sw_no, self.sw_x = to_switch(pin)
        elif socket is not None:
            self.sw_no, self.sw_x = to_switch(to_pin(socket))
        elif (sw_no is not None) and (sw_x is not None):
            self.sw_no, self.sw_x = sw_no, sw_x
        else:
            __logger.error(f'Invalid port information.')

        self.sw_no = -1
        self.sw_x = -1

    def pin(self):
        return (self.sw_no//3)*28 + (self.sw_no%2)*16 + self.sw_x

    def socket(self):
        return to_socket(self.pin())

    def switch(self):
        return self.sw_no, self.sw_x

    def __repr__(self) -> str:
        return f'Pin-{self.pin()}'
